### ISSUE TITLE
[Client] Run comodoo image on kiosk mode

### ISSUE THREAD

**Issue by indisoluble** [comodoo/meta-comodoo#14]
Ideally, after switching on the POS, the only thing the user should see on the screen is the Odoo webpage. To say in another way, right after booting the device, this should launch the web browser on fullscreen mode and block any other action not needed to navigate through the different pages.

**Comment by indisoluble** [comment]
As commented before, Chromium already includes a Kiosk Mode (full screen + block unwanted actions). In the case of Firefox, this same objective has to be reached with an extension. Although [meta-browser](https://layers.openembedded.org/layerindex/branch/master/layer/meta-browser/) has a recipe that does this, called [firefox-addon-webconverger](https://layers.openembedded.org/layerindex/recipe/1767/), it does not work with our image.

I am going to look for another extension and see if I am able to include it in our image.

**Comment by indisoluble** [comment]
Never mind, there are a few extensions that we could use to launch Firefox in kiosk mode but the code is not easily available and each one has a different license.

Now, chromium has its own problems. Although [meta-browser](https://layers.openembedded.org/layerindex/branch/master/layer/meta-browser/) has a branch named [krogoth](https://github.com/OSSystems/meta-browser/tree/krogoth), it is not prepared to compile chromium using this release. I am going to clone all the repos again and checkout [jethro](http://git.yoctoproject.org/cgit/cgit.cgi/poky/log/?h=jethro) and see what happens.

**Comment by indisoluble** [comment]
Well, I know now that I have to checkout jethro but not in all repositories, more exactly I must not checkout [jethro in meta-raspberry](http://git.yoctoproject.org/cgit/cgit.cgi/meta-raspberrypi/log/?h=jethro) because this branch does not include the BSP for RaspberryPi Model 3, only for models 1 & 2. On a first instance, I thought that the BSP for RaspberryPi 3 only included the WIFI driver and all the rest was the same as RaspberryPi 2. Based on this, I spent a few hours generating a image for RaspberryPi 2 and when I inserted it in my RaspberryPi 3, of course, it did not work.

Once I fixed this problem (i.e. after I checkout master for meta-raspberrypi), I tried to generate a new image for RaspberryPi 3 based on jethro with chromium. The problem now is that bitbake is not able to compile chromium because there is not enough memory in my host computer. I am trying now to increase the size of the swap partition.

**Comment by indisoluble** [comment]
Increasing the swap space let BitBake to run, however it is not able to do much because the computer is constantly swapping pages ... which, by the way, it is expected and I should have realised before.

For the record, [these are the minimum specs to compile chromium in Linux](https://chromium.googlesource.com/chromium/src/+/master/docs/linux_build_instructions.md):
> 64-bits: x86 builds are not supported on Linux.
> Memory: >16GB is highly recommended.
> Disk space: Expect a full checkout and build to take nearly 100GB.
> Distribution: You should be able to build Chromium on any reasonably modern Linux distribution, but there are a lot of distributions and we sometimes break things on one or another. Internally, our development platform has been a variant of Ubuntu 14.04 (Trusty Tahr); we expect you will have the most luck on this platform.

**Comment by indisoluble** [comment]
Things are not going well, let me summarise the current status:

* Midori: It is not even worth to try generate an image with this browser because it is not able to load the odoo pages; to be more precise, the version 0.5.11 installed in my host computer eventually crashes trying to load the page.
* Epiphany: It fails with the following error:

```
/path/to/yocto/build/tmp/work/cortexa7hf-neon-vfpv4-poky-linux-gnueabi/webkitgtk/2.10.7-r0/webkitgtk-2.10.7/Source/WebCore/platform/graphics/GLContext.cpp:135:83: error: invalid conversion from 'GLNativeWindowType {aka long long unsigned int}' to 'EGLNativeWindowType {aka void*}' [-fpermissive]
     if (auto eglContext = GLContextEGL::createContext(windowHandle, sharingContext))
                                                                                   ^
In file included from /path/to/yocto/build/tmp/work/cortexa7hf-neon-vfpv4-poky-linux-gnueabi/webkitgtk/2.10.7-r0/webkitgtk-2.10.7/Source/WebCore/platform/graphics/GLContext.cpp:29:0:
/path/to/yocto/build/tmp/work/cortexa7hf-neon-vfpv4-poky-linux-gnueabi/webkitgtk/2.10.7-r0/webkitgtk-2.10.7/Source/WebCore/platform/graphics/egl/GLContextEGL.h:38:42: note:   initializing argument 1 of 'static std::unique_ptr<WebCore::GLContextEGL> WebCore::GLContextEGL::createContext(EGLNativeWindowType, WebCore::GLContext*)'
     static std::unique_ptr<GLContextEGL> createContext(EGLNativeWindowType, GLContext* sharingContext = 0);
                                          ^
[ 62%] Building CXX object Source/WebCore/CMakeFiles/WebCore.dir/platform/graphics/GraphicsContext3DPrivate.cpp.o
Source/WebCore/CMakeFiles/WebCore.dir/build.make:58510: recipe for target 'Source/WebCore/CMakeFiles/WebCore.dir/platform/graphics/GLContext.cpp.o' failed
make[2]: *** [Source/WebCore/CMakeFiles/WebCore.dir/platform/graphics/GLContext.cpp.o] Error 1
make[2]: *** Waiting for unfinished jobs....
cd /path/to/yocto/build/tmp/work/cortexa7hf-neon-vfpv4-poky-linux-gnueabi/webkitgtk/2.10.7-r0/build/Source/WebCore &&/path/to/yocto/build/tmp/sysroots/x86_64-linux/usr/bin/arm-poky-linux-gnueabi/arm-poky-linux-gnueabi-g++   -DBUILDING_GTK__=1 -DBUILDING_WEBKIT -DBUILDING_WITH_CMAKE=1 -DBUILDING_WebCore -DDATA_DIR=\"/usr/share\" -DGETTEXT_PACKAGE=\"WebKit2GTK-4.0\" -DHAVE_CONFIG_H=1 -DMESA_EGL_NO_X11_HEADERS -DUSER_AGENT_GTK_MAJOR_VERSION=602 -DUSER_AGENT_GTK_MINOR_VERSION=1 -DUSE_VCHIQ_ARM -DWEBKITGTK_API_VERSION_STRING=\"4.0\" -DWEBKIT_DOM_USE_UNSTABLE_API -I/path/to/yocto/build/tmp/work/cortexa7hf-neon-vfpv4-poky-linux-gnueabi/webkitgtk/2.10.7-r0/webkitgtk-2.10.7/Source/WebCore -I/path/to/yocto/build/tmp/work/cortexa7hf-neon-vfpv4-poky-linux-gnueabi/webkitgtk/2.10.7-r0/webkitgtk-2.10.7/Source/WebCore/Modules/airplay ... -fPIC  -o CMakeFiles/WebCore.dir/platform/graphics/GraphicsContext3DPrivate.cpp.o -c /path/to/yocto/build/tmp/work/cortexa7hf-neon-vfpv4-poky-linux-gnueabi/webkitgtk/2.10.7-r0/webkitgtk-2.10.7/Source/WebCore/platform/graphics/GraphicsContext3DPrivate.cpp
make[2]: Leaving directory '/path/to/yocto/build/tmp/work/cortexa7hf-neon-vfpv4-poky-linux-gnueabi/webkitgtk/2.10.7-r0/build'
CMakeFiles/Makefile2:607: recipe for target 'Source/WebCore/CMakeFiles/WebCore.dir/all' failed
make[1]: *** [Source/WebCore/CMakeFiles/WebCore.dir/all] Error 2
make[1]: Leaving directory '/path/to/yocto/build/tmp/work/cortexa7hf-neon-vfpv4-poky-linux-gnueabi/webkitgtk/2.10.7-r0/build'
Makefile:152: recipe for target 'all' failed
make: *** [all] Error 2
ERROR: oe_runmake failed
ERROR: Function failed: do_compile (log file is located at/path/to/yocto/build/tmp/work/cortexa7hf-neon-vfpv4-poky-linux-gnueabi/webkitgtk/2.10.7-r0/temp/log.do_compile.9545)
```

* Chromium: So far, it looks like the most complete option, however I am not able to compile it with my host computer, it does not have enough RAM memory.
* Firefox: Complete and easy to compile but it is not possible to launch it in full screen mode unless we add an extension to it. [The extension in meta-browser](https://github.com/OSSystems/meta-browser/tree/master/recipes-mozilla/firefox-addon) does not work in our RaspberryPi image. Other extensions are available but their source code is not available and/or the license does not fit our needs.

As I see it, we have 2 options:

* Try to compile Chromium in a computer with better specs.
* Develop our own extension for Firefox.

**Comment by indisoluble** [comment]
Tested in Fedora24

```
sudo dnf install pywebkitgtk -y
```
Example program (Python)

```
#!/usr/bin/env python

import gtk, webkit

w = gtk.Window()
w.set_title("Example Editor")
w.connect("destroy", gtk.main_quit)
w.resize(800,600)

scroll = gtk.ScrolledWindow()
w.add(scroll)

browser = webkit.WebView()
browser.get_settings().set_property("enable-webgl", True)
scroll.add(browser)

w.show_all()

# browser.open("http://lo-th.github.io/Oimo.js/index.html");
browser.open("https://www.odoo.com/");

gtk.main()

```

**Comment by indisoluble** [comment]
In order to run an embedded application (full screen and so on) for HTML rendering, we need Python, webkit engine and the Python module for webkit (take a look to the example). 

We can take as foundation this meta layer: https://github.com/Igalia/meta-webkit

**Comment by indisoluble** [comment]
Cool, let give this a try. But keep in mind that the error I pasted above for Epiphany is, in fact, an error got while compiling webkitgtk.

**Comment by indisoluble** [comment]
* *Bad news:* There is an issue in [meta-webkit repo](https://github.com/Igalia/meta-webkit/issues/1) that  informs that it is not possible to compile the layer for Raspberry Pi. I will add my own comment ASAP.
* *Good news:* I managed to build an image with [WebKit for Wayland](https://github.com/Igalia/meta-webkit#building-the-webkitforwayland-engine).
* *Bad news:* Wayland is a replacement for X11 and x11vnc does not work with it, to say in another way, I can not see the graphic interface.

**Comment by indisoluble** [comment]
I finally managed to build an image with WebKitGTK. I followed the instructions in [Igalia/meta-webkit](https://github.com/Igalia/meta-webkit) plus some adjustments of my own:

* First of all, I was not able to build an image based on the last release named [krogoth](http://git.yoctoproject.org/cgit/cgit.cgi/poky/log/?h=krogoth), instead I had to use the previous one called [jethro](http://git.yoctoproject.org/cgit/cgit.cgi/poky/log/?h=jethro).
* To be more specific, I built the image with the following configuration:
```
Build Configuration:
BB_VERSION        = "1.28.0"
BUILD_SYS         = "x86_64-linux"
NATIVELSBSTRING   = "Ubuntu-16.04"
TARGET_SYS        = "arm-poky-linux-gnueabi"
MACHINE           = "raspberrypi3"
DISTRO            = "poky"
DISTRO_VERSION    = "2.0.2"
TUNE_FEATURES     = "arm armv7a vfp thumb neon callconvention-hard vfpv4 cortexa7"
TARGET_FPU        = "vfp-vfpv4-neon"
meta              
meta-yocto        
meta-yocto-bsp    = "jethro:40376446904ae3529be41737fed9a0b650ed167d"
meta-raspberrypi  = "master:cc64d6324d1543f009f0c010d720e74b7cb9f5d5"
meta-webkit       = "jethro:a107c901f93d453f886a3014224ca2430d15fcf3"
```

* Also, it was not enough to add the following block to `local.conf`:
```
DISTRO_FEATURES_append = " opengl x11"
IMAGE_INSTALL_append = " webkitgtk-bin"
```
I ended up adding these lines to `local.conf`:
```
MACHINE ?= "raspberrypi3"
DISTRO_FEATURES_append = " opengl x11"
IMAGE_INSTALL_append = " webkitgtk-bin"
PREFERRED_PROVIDER_virtual/libgl = "mesa"
PREFERRED_PROVIDER_virtual/libgles2 = "mesa"
PREFERRED_PROVIDER_virtual/egl = "mesa"
PREFERRED_PROVIDER_virtual/mesa = "mesa"
```

I also had to increase the size of the swap space ... but that is another history.

The next steps are:

1. Report my findings in the [Igalia/meta-webkit repo](https://github.com/Igalia/meta-webkit), more exactly in the [issue where someone commented that it was not possible to compile this engine in RaspberryPi](https://github.com/Igalia/meta-webkit/issues/1).
2. Add the GTK and WebKit python modules to our image so we can run the script pasted above.

**Comment by indisoluble** [comment]
Done, [comment is in place](https://github.com/Igalia/meta-webkit/issues/1#issuecomment-245689674).

**Comment by indisoluble** [comment]
I was doing some tests with pywebkitgtk package, and the Odoo login page breaks the application. Probably pywebkitgtk binding is not very well maintained or updated right now.

On the other hand, I was doing the same tests with python-qt5-webkit. QT is a well maintained toolkit, and a huge user community around it. Nevertheless, we have similar problems: The Odoo login page loads with a few "client side" errors.

Again the problem is related to the bindings (the Python bindings). 

We have to take a decision, try to fix the warnings in QT5 webkit program, or try things like this: https://dzone.com/articles/what-is-nwjs

**Comment by indisoluble** [comment]
```
#!/usr/bin/env python
from PyQt5.QtCore import QUrl
from PyQt5.QtWidgets import QApplication
from PyQt5.QtWebKitWidgets import QWebView
import sys

app = QApplication(sys.argv)
view = QWebView()
view.setUrl(QUrl("http://javiroman.odoo.com"))
view.show()
sys.exit(app.exec_())
```

**Comment by indisoluble** [comment]
Try something like: https://dzone.com/articles/what-is-nwjs, it is too far from my reach at this moment, I am afraid.

About QT5 ... well, there is already a layer available: https://github.com/meta-qt5/meta-qt5, which makes things a lot easier for me.

Also, I was thinking on using Gecko (Firefox engine) as we intended to use WebKit given that we were able to load and use Odoo webpage with this engine. I found this GitHub repo: https://github.com/mozilla-japan/gecko-embedded, and it seems they are doing some job in this regard.

**Comment by indisoluble** [comment]
Anyway, right now I am building a new image with WebKitGTK, one last try to see if I am able to load and show Odoo POS webpage. If not, I will try to build a new image, this time with QT5; and we will see what we can do from this point.

BTW, we know Gecko is able to render the Odoo webpage, however I do not know how we can install just the engine and not the entire browser. But, is this really a problem? As far as I know, **our main problem with Firefox was that it required an extension** to present it in kiosk mode, however we could develop our own. After all the work done so far, I tend to think that this option is safer and can be done in less time. So, **why is it better to install just the engine plus the necessary python packages?** Will it make things easier when we try to integrate the barcode reader and the rest of the hardware?

**Comment by indisoluble** [comment]
Just to be sure we all are in the same page, let me summarize the current status:
1. I could not get any results with WebKitGTK, so I discarded it and started to work on QT5.
2. At the moment of writing this text, I already have an poky image with QT5.
3. Sadly, the demo browser included in the image is not able to load the POS webpage (see image attached with title _QT5BrowserError.png_).
4. Also, I have not been able to install PyQt5 but I realised Yocto already has a recipe to include PyQt4. Given that QT5 fails, I am going to try to build an image with QT4 and see what happens.

Let me know what you think @javiroman . Also, I am still waiting for an answer to my question in the comment above :cry:

**Comment by indisoluble** [comment]
Going over all this comments, probably the cleverest is:

"BTW, we know Gecko is able to render the Odoo webpage, however I do not know how we can install just the engine and not the entire browser. But, is this really a problem? As far as I know, our main problem with Firefox was that it required an extension to present it in kiosk mode, however we could develop our own."

Apparently this is the only via, for our requirements (Odoo, nodejs requirements).

**Comment by indisoluble** [comment]
Well, as expected, we can not load the POS webpage with QT4 either (see image _QT4BrowserError.png_).

Let go back to Firefox and see what we can do with it.

**Comment by indisoluble** [comment]
I finally managed to prepare an image able to launch Firefox on kiosk mode (GitHub branch: [krogoth-firefox](https://github.com/indisoluble/meta-comodoo/tree/krogoth-firefox)). However, the browser is not launched automatically, I think this is enough for the moment and before going any further on this direction, we should consider other tasks like interfacing with the rest of the hardware. Anyway, the steps to launch the kiosk are:

1. Switch on the RaspberryPi
2. Connect by ssh to the RaspberryPi
3. On the ssh terminal, launch `x11vnc &`
4. When the VNC server is ready, open a VNC client on your host computer
5. Back on the ssh terminal, execute `comodoo.sh &`
6. On the VNC client windows, Firefox will appear on kiosk mode and load the Odoo login page.

To end, I did not develop our own extension, instead of that I have used the addon named [R-kiosk](https://addons.mozilla.org/en-us/firefox/addon/r-kiosk/), which has a permissive enough license and also the code is available [here](https://github.com/un1xoid/r-kiosk).
